app/models/user.rb

```diff
@@ -0,0 +1,31 @@
+class User < ApplicationRecord
+    has_many :products
+    has_many :orders
+    has_many :order_items
+    validates :username, presence: true, uniqueness: true
+    validates :email, presence: true, uniqueness: true
+
+
+    def self.build_from_github(auth_hash)
+        user = User.new
+        user.uid = auth_hash[:uid]
+        user.provider = "github"
+        user.username = auth_hash["info"]["name"]
```

Comment:
This line causes the login bug that I mentioned in the feedback notes.  In the auth_hash ["info"]["name"] is an optional field and can be nil.  If it is nil (as mine was when I tested) this causes the validation that the username has to exist to fail.  To fix this, I would suggest using auth_hash["info"]["login"] as an alternative to name, or removing the validation.  It is important to check that if you are validating on something, it will always exist in a valid case.